flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismAndInputInfosDeciderTest.java

```diff
@@ -379,6 +379,30 @@ void testEvenlyDistributeDataWithMaxSubpartitionLimitation() {
                 subpartitionRanges);
     }
 
+    @Test
+    void testComputeSourceParallelismUpperBound() {
+        Configuration configuration = new Configuration();
+        // Test the case where default source parallelism is not configured, falling back to global
+        // max parallelism.
+        VertexParallelismAndInputInfosDecider vertexParallelismAndInputInfosDecider =
+                DefaultVertexParallelismAndInputInfosDecider.from(32, configuration);
+        assertThat(
+                        vertexParallelismAndInputInfosDecider.computeSourceParallelismUpperBound(
+                                new JobVertexID(), 128))
+                .isEqualTo(32);
+
+        // Test the case where the configured default source parallelism cannot exceed the vertex
+        // max parallelism.
+        configuration.setInteger(
```

Comment:
It's better to create different tests for different cases, e.g. `testComputeSourceParallelismUpperBound` for normal case, `testComputeSourceParallelismUpperBoundFallback` for the fallback case, `testComputeSourceParallelismUpperBoundNotExceedMaxParallelism` for verifying max parallelism effectiveness.